core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -482,21 +482,28 @@ private[spark] class TaskSchedulerImpl(
     }
   }
 
-  /** Remove an executor from all our data structures and mark it as lost */
+  /**
+   * Remove an executor from all our data structures and mark it as lost. If the executor's loss
+   * reason is not yet known, do not yet remove its association with its host.
+   */
   private def removeExecutor(executorId: String, reason: ExecutorLossReason) {
     activeExecutorIds -= executorId
     val host = executorIdToHost(executorId)
     val execs = executorsByHost.getOrElse(host, new HashSet)
-    execs -= executorId
-    if (execs.isEmpty) {
-      executorsByHost -= host
-      for (rack <- getRackForHost(host); hosts <- hostsByRack.get(rack)) {
-        hosts -= host
-        if (hosts.isEmpty) {
-          hostsByRack -= rack
+
+    if (reason != LossReasonPending) {
```

Comment:
I can't remove the information about where the executor was running; otherwise, the second time `executorLost` is called, `host` (L491) will be null.

I could change the code so that only `executorsByHost -= host` is within the check, but it didn't seem necessary. What really controls whether things are scheduled on the executor is `CoarseGrainedSchedulerBackend` calling `resourceOffers`, and it won't offer resources for executors that are pending removal.